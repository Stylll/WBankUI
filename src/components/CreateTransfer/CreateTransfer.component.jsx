import React, { useState, useEffect, useRef } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as accountActions from '../../redux/actionCreators/accountActions';
import TextInput from '../TextInput/TextInput.component';
import Button from '../Button/Button.component';
import Dropdown from '../Dropdown/Dropdown.component';
import { validateCreateTransfer } from '../../helpers/validators/accountValidator';
import { currencies } from '../../helpers/defaults';
import './CreateTransfer.styles.scss';

const CreateTransfer = ({
  createTransfer, isLoading: reduxIsLoading, closeModal,
  errorMessage, accountTransferSuccess, dataTestId,
}) => {
  const [values, setValues] = useState({
    accountNo: '',
    transferAccountNo: '',
    amount: '',
    currency: '',
  });

  const [formErrors, setFormErrors] = useState({
    accountNo: '',
    transferAccountNo: '',
    amount: '',
    currency: '',
  });

  const [isLoading, setIsLoading] = useState(false);

  const [serverError, setServerError] = useState('');

  const formSubmitted = useRef(false);
  const hasServerOperationOccurred = useRef(false);

  useEffect(() => {
    if (!reduxIsLoading && accountTransferSuccess && formSubmitted.current) {
      setIsLoading(false);
      closeModal();
      formSubmitted.current = false;
      return;
    }

    if (!reduxIsLoading && formSubmitted.current) {
      setIsLoading(false);
      formSubmitted.current = false;
    }
  }, [reduxIsLoading, accountTransferSuccess, closeModal]);

  useEffect(() => {
    if (errorMessage && hasServerOperationOccurred.current) {
      setServerError(errorMessage);
    }
  }, [errorMessage]);

  const handleChange = (event) => {
    event.preventDefault();
    setValues({
      ...values,
      [event.target.name]: event.target.value,
    });
  };

  const handleClick = (event) => {
    event.preventDefault();
    const { isValid, errors } = validateCreateTransfer(values);
    setFormErrors({
      ...errors,
    });
    if (!isValid) {
      return;
    }

    setIsLoading(true);

    createTransfer({
      accountNo: values.accountNo,
      transferAccountNo: values.transferAccountNo,
      amount: values.amount,
      currency: values.currency,
    });

    formSubmitted.current = true;
    hasServerOperationOccurred.current = true;
  };

  const renderCreateForm = () => (
    <div className="transfer-container">
        <div>
            <p>Make a transfer</p>
            <span className="error-text">{serverError}</span>
            <TextInput
                name="accountNo"
                placeholder="3245"
                title="Enter the account number to transfer from"
                type="text"
                value={values.accountNo}
                error={formErrors.accountNo}
                onChange={handleChange}
                dataTestId={`${dataTestId}-accountNo`} />
            <TextInput
                name="transferAccountNo"
                placeholder="1294"
                title="Enter the account number to transfer to"
                type="text"
                value={values.transferAccountNo}
                error={formErrors.transferAccountNo}
                onChange={handleChange}
                dataTestId={`${dataTestId}-transferAccountNo`} />
            <TextInput
                name="amount"
                placeholder="5000"
                title="Enter the amount to transfer"
                type="text"
                value={values.amount}
                error={formErrors.amount}
                onChange={handleChange}
                dataTestId={`${dataTestId}-amount`} />
            <Dropdown
                options={currencies}
                name="currency"
                optionsTitleProperty="name"
                optionsValueProperty="id"
                value={values.currency}
                error={formErrors.currency}
                onChange={handleChange}
                placeholder="select amount currency"
                dataTestId={`${dataTestId}-currency`}
            />
            <br />
            <Button
                title="Transfer"
                handleClick={handleClick}
                showLoader={isLoading}
                disabled={isLoading}
                dataTestId={`${dataTestId}-transfer`}
            />
        </div>
        <br />
    </div>
  );

  return (
        <div className="transfer-account-container">
            <div className="form-container">
                    {renderCreateForm()}
                </div>
        </div>
  );
};

CreateTransfer.propTypes = {
  createTransfer: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  accountTransferSuccess: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string.isRequired,
  dataTestId: PropTypes.string,
};

CreateTransfer.defaultProps = {
  dataTestId: 'createtransfer',
};

const mapStateToProps = ({ account }) => ({
  isLoading: account.isTransferAccountLoading,
  accountTransferSuccess: account.accountTransferSuccess,
  errorMessage: account.errorMessage,
});

const mapDispatchToProps = {
  createTransfer: accountActions.createTransfer,
};

export {
  CreateTransfer,
};

export default connect(mapStateToProps, mapDispatchToProps)(CreateTransfer);
