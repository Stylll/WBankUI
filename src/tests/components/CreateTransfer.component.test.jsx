import { render, fireEvent } from '@testing-library/react';
import React from 'react';

import { CreateTransfer } from '../../components/CreateTransfer/CreateTransfer.component';

describe('CreateTransfer Test', () => {
  it('should render properly', () => {
    const { baseElement } = render(<CreateTransfer
      createTransfer={() => {}}
      closeModal={() => {}}
      isLoading={false}
      accountTransferSuccess={false}
      errorMessage=""
    />);
    expect(baseElement).toMatchSnapshot();
  });
  it('should validate the input', () => {
    const { baseElement, getByTestId, getByText } = render(<CreateTransfer
      createTransfer={() => {}}
      closeModal={() => {}}
      isLoading={false}
      accountTransferSuccess={false}
      errorMessage=""
    />);
    fireEvent.click(getByTestId('createtransfer-transfer'));
    expect(getByText('account number cannot be empty')).toBeInTheDocument();
    expect(getByText('transfer account number cannot be empty')).toBeInTheDocument();
    expect(getByText('amount cannot be empty')).toBeInTheDocument();
    expect(getByText('currency must be either cad, usd, or mxn')).toBeInTheDocument();
  });

  it('should submit form, call createTransfer and display error', () => {
    const createTransfer = jest.fn();
    const {
      baseElement, getByTestId, getByText, rerender,
    } = render(<CreateTransfer
      createTransfer={createTransfer}
      closeModal={() => {}}
      isLoading={false}
      accountTransferSuccess={false}
      errorMessage=""
    />);

    fireEvent.change(getByTestId('createtransfer-accountNo'),
      { target: { value: '2345', name: 'accountNo' } });
    fireEvent.change(getByTestId('createtransfer-transferAccountNo'),
      { target: { value: '1234', name: 'transferAccountNo' } });
    fireEvent.change(getByTestId('createtransfer-amount'),
      { target: { value: '245', name: 'amount' } });
    fireEvent.change(getByTestId('createtransfer-currency'),
      { target: { value: 'usd', name: 'currency' } });
    fireEvent.click(getByTestId('createtransfer-transfer'));
    expect(createTransfer).toHaveBeenCalled();

    rerender(<CreateTransfer
      createTransfer={createTransfer}
      closeModal={() => {}}
      isLoading={false}
      accountTransferSuccess={false}
      errorMessage="An error occurred"
    />);
    expect(getByText('An error occurred')).toBeInTheDocument();
  });
});
