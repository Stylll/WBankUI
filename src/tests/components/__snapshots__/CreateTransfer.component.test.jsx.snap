// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTransfer Test should render properly 1`] = `
<body>
  <div>
    <div
      class="transfer-account-container"
    >
      <div
        class="form-container"
      >
        <div
          class="transfer-container"
        >
          <div>
            <p>
              Make a transfer
            </p>
            <span
              class="error-text"
            />
            <div
              class="form-input"
            >
              <div
                class="title"
              >
                Enter the account number to transfer from
              </div>
              <input
                class="text-input"
                data-testid="createtransfer-accountNo"
                id=""
                name="accountNo"
                placeholder="3245"
                type="text"
                value=""
              />
              <span
                class="info"
              />
              
            </div>
            <div
              class="form-input"
            >
              <div
                class="title"
              >
                Enter the account number to transfer to
              </div>
              <input
                class="text-input"
                data-testid="createtransfer-transferAccountNo"
                id=""
                name="transferAccountNo"
                placeholder="1294"
                type="text"
                value=""
              />
              <span
                class="info"
              />
              
            </div>
            <div
              class="form-input"
            >
              <div
                class="title"
              >
                Enter the amount to transfer
              </div>
              <input
                class="text-input"
                data-testid="createtransfer-amount"
                id=""
                name="amount"
                placeholder="5000"
                type="text"
                value=""
              />
              <span
                class="info"
              />
              
            </div>
            <div
              class="drop-down"
            >
              <select
                class="select"
                data-testid="createtransfer-currency"
                name="currency"
              >
                <option
                  value=""
                >
                  select amount currency
                </option>
                <option
                  value="cad"
                >
                  CAD
                </option>
                <option
                  value="usd"
                >
                  USD
                </option>
                <option
                  value="mxn"
                >
                  MXN
                </option>
              </select>
              <span
                class="chevy"
              >
                âˆ¨
              </span>
              <span
                class="info"
              />
              
            </div>
            <br />
            <div
              class="btn-button"
            >
              <button
                class="button "
                data-testid="createtransfer-transfer"
                type="submit"
              >
                Transfer
              </button>
            </div>
          </div>
          <br />
        </div>
      </div>
    </div>
  </div>
</body>
`;
